### PR TITLE
Add the unstack operation to keras core

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
I recently got a use-case for the [`unstack`](https://www.tensorflow.org/api_docs/python/tf/unstack) operation while working with the [segment anything model tensorflow port](https://github.com/tirthasheshpatel/segment_anything_tensorflow/blob/b93d715e96ad6838541972debfac76e56d1f6bfe/sam_tf/image_encoder.py#L105). This PR adds the op.

Note that, if the input tensor is symbolic, it is not straight-forward to compute the output spec since the output list length might be arbitrary. So, this PR only adds support for non-symbolic tensors for the time being. Support for symbolic tensors can be added as a follow-up enhancement.

cc @ianstenbit

### CODE REVIEW

**Comment by fchollet** [comment]
I am not sure we should add this operation --

1. All our ops are both eager and symbolic, so it would be inconsistent.
2. It's not in the NumPy API (whereas `stack` is) and not in JAX / torch

**Comment by tirthasheshpatel** [comment]
> * All our ops are both eager and symbolic, so it would be inconsistent.

I will add a symbolic version of this op. Initially, I thought that keras is inferring the shape even if it is `None` in the axis which the user wishes to unstack. But this is not the case: Keras just errors out:

```python
```

**Comment by tirthasheshpatel** [comment]
> 2\. It's not in the NumPy API (whereas `stack` is) and not in JAX / torch

I agree, but there is no way to replicate the op in TensorFlow or PyTorch without using the unstack operation i.e. tensorflow as torch don't allow just unpacking a tensor like `a, b, c = ops.random.normal((3, 4))`. Also, it is very easy to implement the op in NumPy and Jax. So, I am more on the side of adding it than not.

**Comment by fchollet** [comment]
> I will add a symbolic version of this op. Initially, I thought that keras is inferring the shape even if it is None in the axis which the user wishes to unstack. But this is not the case: Keras just errors out:

I guess you could have symbolic support but require a defined shape?

**Comment by tirthasheshpatel** [comment]
> I guess you could have symbolic support but require a defined shape?

Yes, exactly.

File: keras_core/ops/core.py [hunk 1]
```diff
@@ -344,6 +344,58 @@ def fori_loop(lower, upper, body_fun, init_val):
...
+        ]
+        return output
+
+
+@keras_core_export("keras_core.ops.unstack")
+def unstack(x, num=None, axis=0):
+    """Unpacks the given dimension of a rank-R tensor into rank-(R-1) tensors.
+
+    Args:
+        x: The input tensor (must be non-symbolic).
```
Comment (tirthasheshpatel) [inline keras_core/ops/core.py:382]:
```suggestion
        x: The input tensor.
```

File: keras_core/ops/core.py [hunk 2]
```diff
@@ -344,6 +344,58 @@ def fori_loop(lower, upper, body_fun, init_val):
...
+        axis: The axis along which to unpack.
```
Comment (tirthasheshpatel) [inline keras_core/ops/core.py:383]:
```suggestion
        num: The length of the dimension axis. Automatically inferred
            if `None`.
        axis: The axis along which to unpack.
```

File: keras_core/ops/core_test.py [hunk 3]
```diff
@@ -56,6 +56,23 @@ def body_fun(i, x):
...
+            self.assertEqual(o.shape, (2, 4))
+
+        x = KerasTensor((2, None, None))
+        axis, num = 1, 3
+        out = core.unstack(x, num=num, axis=axis)
+        self.assertEqual(len(out), 3)
+        for o in out:
+            self.assertEqual(o.shape, (2, None))
+
+        pytest.raises(ValueError, lambda: core.unstack(x, axis=axis))
```
Comment (fchollet) [inline keras_core/ops/core_test.py:74]: Use `self.assertRaisesRegex` and check (part of) the error message

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
